src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala

```diff
@@ -265,43 +267,82 @@ class DeepDiveLogParser extends JavaTokenParsers {
                            ~ "returns" ~ outTy
-                       ~ implementationDecls ~ mode) =>
-             FunctionDeclaration(a, inTy, outTy, implementationDecls, mode.getOrElse(null))
+                       ~ implementationDecls) =>
+             FunctionDeclaration(a, inTy, outTy, implementationDecls, mode)
     }
 
-  def extractionRule : Parser[ExtractionRule] =
-    conjunctiveQuery ~ opt(supervision) ^^ {
-      case (q ~ supervision) => ExtractionRule(q, supervision)
-    }
+  def cqBody: Parser[Body] = cond | quantifiedBody | atom
+
+  def cqConjunctiveBody: Parser[List[Body]] = rep1sep(cqBody, ",")
+
+  def cqHeadTerms = "(" ~> rep1sep(expr, ",") <~ ")"
+
+  def conjunctiveQuery : Parser[ConjunctiveQuery] =
+    cqHeadTerms ~ opt("*") ~ opt("|" ~> decimalNumber) ~ ":-" ~ rep1sep(cqConjunctiveBody, ";") ^^ {
+      case (head ~ isDistinct ~ limit ~ ":-" ~ disjunctiveBodies) =>
+        ConjunctiveQuery(head, disjunctiveBodies, isDistinct != None, limit map (_.toInt))
+  }
 
   def functionCallRule : Parser[FunctionCallRule] =
-    ( relationName ~ ":-" ~ "!"
-    ~ functionName ~ "(" ~ relationName ~ ")"
-    ) ^^ {
-      case (out ~ ":-" ~ "!" ~ fn ~ "(" ~ in ~ ")") =>
-        FunctionCallRule(in, out, fn)
+    relationName ~ "+=" ~ functionName ~ conjunctiveQuery ^^ {
+      case (out ~ _ ~ func ~ cq) => FunctionCallRule(out, func, cq)
     }
 
-  def factorWeight = "weight" ~> "=" ~> rep1sep(expr, ",") ^^ { FactorWeight(_) }
+  def supervision = "@label" ~> "(" ~> variableName <~ ")"
+  def extractionRule =
+    opt(supervision) ~ relationName ~ conjunctiveQuery ^^ {
+      case (sup ~ head ~ cq) => ExtractionRule(head, cq, sup)
+  }
 
-  def supervision = "label" ~> "=" ~> variableName
+  def factorWeight = "@weight" ~> "(" ~> rep1sep(expr, ",") <~ ")" ^^ { FactorWeight(_) }
+  def inferenceMode = "@mode" ~> "(" ~> inferenceModeType <~ ")"
 
-  def factorFunctionName = "Imply" | "And" | "Equal" | "Or" | "Multinomial" | "Linear" | "Ratio"
-  def factorFunction = "function" ~> "=" ~> factorFunctionName
+  // factor functions
+  def headAtom = relationName ~ ("(" ~> rep1sep(expr, ",") <~ ")") ^^ {
+    case (r ~ expressions) => HeadAtom(r, expressions)
+  }
+  def implyHead = rep1sep(headAtom, ",") ~ "=>" ~ headAtom ^^ {
+    case (a ~ _ ~ b) => InferenceRuleHead(FactorFunction.Imply, a :+ b)
+  }
+  def isTrueHead = headAtom ^^ { x => InferenceRuleHead(FactorFunction.IsTrue, List(x)) }
+  def equalHead = headAtom ~ "=" ~ rep1sep(headAtom, "=") ^^ { case (a ~ _ ~ b) =>
+    InferenceRuleHead(FactorFunction.Equal, a +: b)
+  }
+  def andHead   = headAtom ~ "^" ~ rep1sep(headAtom, "^") ^^ { case (a ~ _ ~ b) =>
+    InferenceRuleHead(FactorFunction.And, a +: b)
+  }
+  def orHead    = headAtom ~ "v" ~ rep1sep(headAtom, "v") ^^ { case (a ~ _ ~ b) =>
+    InferenceRuleHead(FactorFunction.Or, a +: b)
+  }
+  def multinomialHead = "Multinomial" ~> "(" ~> rep1sep(headAtom, ",") <~ ")" ^^ {
```

Comment:
I didn't manage to come up with a nice symbol... so I used the function name here